projects/illm/gan_compression_module.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import hydra
+import torch
+import torch.nn as nn
+from omegaconf import DictConfig
+from target_compression_module import TargetRateCompressionModule
+from torch import Tensor
+
+from neuralcompression import HyperpriorOutput
+from neuralcompression.loss_fn import (
+    DiscriminatorLoss,
+    DistortionLoss,
+    GeneratorLoss,
+    TargetRateConfig,
+)
+from neuralcompression.models import HyperpriorAutoencoderBase
+
+
+class GANCompressionModule(TargetRateCompressionModule):
+    def __init__(
+        self,
+        model: HyperpriorAutoencoderBase,
+        target_rate_config: TargetRateConfig,
+        discriminator: nn.Module,
+        discriminator_loss: DiscriminatorLoss,
+        generator_loss: GeneratorLoss,
+        optimizer_config: DictConfig,
+        distortion_lam: float = 1.0,
+        distortion_loss: Optional[DistortionLoss] = None,
+        latent_projector: Optional[nn.Module] = None,
+        generator_weight: float = 1.0,
+        gradient_clip_val: Optional[float] = None,
+        gradient_clip_algorithm: str = "norm",
+        noisy_context_steps: int = 0,
+        mc_sampling: bool = True,
+    ):
+        super().__init__(
+            model=model,
+            target_rate_config=target_rate_config,
+            optimizer_config=optimizer_config,
+            distortion_lam=distortion_lam,
+            distortion_loss=distortion_loss,
+            gradient_clip_val=gradient_clip_val,
+            gradient_clip_algorithm=gradient_clip_algorithm,
+            noisy_context_steps=noisy_context_steps,
+        )
+
+        self.discriminator = discriminator
+        self.discriminator_loss = discriminator_loss
+        self.latent_projector = latent_projector
+        self.generator_loss = generator_loss
+        self.generator_weight = generator_weight
+        self.mc_sampling = mc_sampling
+
+    def disable_gradients(self, opt):
+        for group in opt.param_groups:
+            for param in group["params"]:
+                param.requires_grad_(False)
+
+    def enable_gradients(self, opt):
+        for group in opt.param_groups:
+            for param in group["params"]:
+                param.requires_grad_(True)
+
+    def run_discriminator(self, images: Tensor, context: Tensor) -> Tensor:
+        if self.discriminator.is_conditional:
+            return self.discriminator(images, context)
+        else:
+            return self.discriminator(images)
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        full_images: Tensor
+        output: HyperpriorOutput
```

Comment:
Oh interesting ! You declare the types at the start ?

Reply:
Ah yes, this is just a way to help my poor little VS Code give me the right functions and types for each variable.